### PR TITLE
Dendrite metrics

### PR DESCRIPTION
**Pull request by karangrewal**
In the dendrites folder, `utils.py` now contains various functions for computing dendrite metrics, and `plotting_utils.py` calls on those functions to create visualizations. The metrics introduced are

- **percent active dendrites**: the fractions of instances for which each dendrite segments is selected per class of inputs
- **mean selected activations**: the mean activation value of each dendrite segments over all instances for which said dendrite segment was selected, grouped according to class,
- **dendrite overlap matrix**: a matrix of "overlap values" that tell how how frequently the same dendrites are selected for any pair of classes,
- **dendrite overlap score**: an aggregate statistic of the overlap matrix for a given set of dendrite segments,
- **duty cycle**: the fraction of total instances for which each dendrite becomes active (similar to percent active dendrites, but computed over an entire dataset instead of class-by-class), and
- **entropy**: the entropy of the observed dendrite activation distribution.

The functions to compute metrics are specifically separate from the plotting functions. To see the visuals created by these new modules, click [here](https://drive.google.com/file/d/1ddXMwcbiW7AN8_QLCSCH81Xuy6x4G4Di/view?usp=sharing).

### CODE REVIEW

File: nupic/research/frameworks/dendrites/plotting_utils.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    dendrite_overlap,
+    dendrite_overlap_matrix,
+    entropy,
+    mean_selected_activations,
+    percent_active_dendrites,
+)
+
+
+def plot_dendrite_activations(
+    dendrite_segments,
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/plotting_utils.py:37]: Looks well documented and clean. Can you make wandb optional? (I'm not familiar enough with wandb to know which lines of code are specific to it.). I should be able to use it from Jupyter lab or just save to a file. Since you are returning a figure, the client can decide what to do with it.

File: nupic/research/frameworks/dendrites/utils.py [hunk 2]
```diff
@@ -19,78 +19,251 @@
...
+    :param context_vectors: iterable of 2D torch tensors with shape (num_examples,
+                            dim_context) where each 2D tensor gives a batch of context
+                            vectors from the same category
+    :param selection_criterion: the criterion for selecting which dendrites become
+                                active; either "regular" (for `GatingDendriticLayer`)
+                                or "absolute" (for `AbsoluteMaxGatingDendriticLayer`)
     """
-    assert dendritic_weights.size(1) == context_vectors.size(1)
-    assert annotation_type in ("regular", "absolute")
+    num_units, num_dendrites, _ = dendrite_segments.weights.size()
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/utils.py:41]: Should there be a `with torch.no_grad():` wrapper around the code below (and also in the other functions)?

**Review by subutai — CHANGES_REQUESTED** [review]

File: nupic/research/frameworks/dendrites/utils.py [hunk 3]
```diff
@@ -19,78 +19,251 @@
...
+    duty_cycle = duty_cycle.sum(axis=0, dtype=torch.float)
+    duty_cycle = duty_cycle / num_examples
+
+    return duty_cycle
+
+
+def online_dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion,
+                               online=False, batch_size=32, period=1000):
+    """
+    Returns a 2D torch tensor with shape (number of units, number of dendrites) where
```
Comment (mvacaporale) [inline nupic/research/frameworks/dendrites/utils.py:199]: Same comment on naming `number of dendrites` vs `number of segments` as above.

File: nupic/research/frameworks/dendrites/utils.py [hunk 4]
```diff
@@ -19,78 +19,251 @@
...
+    ltril_mask = ltril_mask.unsqueeze(0).repeat((num_units, 1, 1))
+
+    overlap_score = (overlap_matrix * ltril_mask).sum(dim=(1, 2))
+    overlap_score /= (0.5 * num_categories * (num_categories - 1))
+    return overlap_score
+
+
+def dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion):
+    """
+    Returns a 2D torch tensor with shape (number of units, number of dendrites) where
```
Comment (mvacaporale) [inline nupic/research/frameworks/dendrites/utils.py:168]: We should be consistent about how we count the dendrites. I've been counting the segments, so the shape here would be `(number of units, number of segments)`. Any thoughts on which might be better?

File: nupic/research/frameworks/dendrites/utils.py [hunk 5]
```diff
@@ -19,78 +19,251 @@
...
+        activations = activations.abs()
+
+    duty_cycle = (activations.max(axis=2, keepdims=True).values == activations)
+    duty_cycle = duty_cycle.sum(axis=0, dtype=torch.float)
+    duty_cycle = duty_cycle / num_examples
+
+    return duty_cycle
+
+
+def online_dendrite_duty_cycle(dendrite_segments, context_vectors, selection_criterion,
```
Comment (mvacaporale) [inline nupic/research/frameworks/dendrites/utils.py:196]: Is this function getting used at all? I don't see any reference to it in `nupic.research`.

File: nupic/research/frameworks/dendrites/utils.py [hunk 6]
```diff
@@ -19,78 +19,251 @@
...
+        batch_start = batch_start + batch_size
+        batch_end = min(num_examples, batch_start + batch_size)
+
+    return duty_cycle
+
+
+def entropy(x):
+    """
+    Returns a tuple of scalars (entropy value, maximum possible entropy value) which
+    gives the entropy of dendrite segments and the maximmum entropy that can be
```
Comment (mvacaporale) [inline nupic/research/frameworks/dendrites/utils.py:254]: "maximum"

**Review by mvacaporale — CHANGES_REQUESTED** [review]
I'm still reviewing, but I some questions that would help me understand things a bit better first. Please see them below.

In general, I believe some changes will be needed to make it easier to plot these metrics throughout training. Here is a general question that may help simplify things...

In `utils.py`, the functions typically take arguments `dendrite_segments`, `context_vectors`, and `selection_criterion`. Suppose you were given instead only the following:
* `dendrite_activations` - these are the outputs of a `dendrite_segments` module of shape `(batch_size, num_units)`
* `winning_mask` - this is a mask of the winning units of shape `(batch_size, num_units, num_segments`); 1 would indicate that segment won for that unit, while 0 would indicate that it lost.

With just these, would you be able to compute all the needed outputs from `utils.py`? I believe this would be the case, but I want to be sure. If so, I believe the same would be true for `plotting_utils.py`, barring additional info such as class labels and which unit to plot.

File: nupic/research/frameworks/dendrites/plotting_utils.py [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+    if category_names is not None:
+        x_labels = category_names
+    y_labels = ["dendrite {}".format(j) for j in range(num_dendrites)]
+
+    percentage_activations = percent_active_dendrites(
+        dendrite_segments=dendrite_segments,
+        context_vectors=context_vectors,
+        selection_criterion=selection_criterion
+    )
+    percentage_activations = percentage_activations[0, :, :].detach().cpu().numpy()
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/plotting_utils.py:155]: Good catch, I believe this is an error, and the user should specify the unit.

File: nupic/research/frameworks/dendrites/utils.py [hunk 8]
```diff
@@ -19,78 +19,251 @@
...
     """
-    assert dendritic_weights.size(1) == context_vectors.size(1)
-    assert annotation_type in ("regular", "absolute")
+    num_units, num_dendrites, _ = dendrite_segments.weights.size()
+
+    percentage_activations = torch.zeros((num_units, num_dendrites, 0))
+    for j in range(len(context_vectors)):
+        num_examples = context_vectors[j].size(0)
+
+        activations = dendrite_segments(context_vectors[j])
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py:47]: Here, the docstring says that `context_vectors` needs to be an "iterable of 2D torch tensors with shape (num_examples, dim_context)", because we want the context vectors separated by class. That means `context_vectors[j]` will be a 2D tensor.

File: nupic/research/frameworks/dendrites/utils.py (hunk 4 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py:168]: We may be referring to the same thing. By "dendrites", I'm referring here to a weight vector (and in this case, each unit has 10 dendrites). The naming is confusing because the `DendriteSegments` class makes it sound like a segment is a set of weights for the entire layer of units. If you propose that we call each weight vector a "segment", then I'm on board with that. In that case, a "dendrite" would just be a single weight scalar in that weight vector.

File: nupic/research/frameworks/dendrites/utils.py (hunk 5 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py:196]: It's not. I only left it in because I thought we may want to consider duty cycle that's not over the entire learning trajectory, and only the last T samples seen. I think I can just merge `dendrite_duty_cycle` and `online_dendrite_duty_cycle` into 1 function with a parameter that controls how duty cycle values are computed.

File: nupic/research/frameworks/dendrites/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/utils.py:47]: I see, missed that, thanks

File: nupic/research/frameworks/dendrites/utils.py (hunk 4 above)
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/utils.py:168]: I believe we were referring to the same thing. My proposal was to call each weight vector a "segment". I think it's less ambiguous this way. It's also more consistent as the `DendritesSegments` module has an attribute called `num_segements`.

File: nupic/research/frameworks/dendrites/utils.py (hunk 5 above)
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/utils.py:196]: In general, we should be wary of added code or functionality that's not being currently used. Code that's unused goes untested and quickly becomes outdated. You should save the function on a separate branch, patch, or stash, and reintroduce it as needed.

**Comment by karangrewal** [comment]
@mvacaporale given just the 2 tensors `dendrite_activations` and `winning_mask` that you propose, I believe all metrics can still be computed.

File: nupic/research/frameworks/dendrites/utils.py (hunk 5 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py]: Ok, I have removed this function from the commit.

File: nupic/research/frameworks/dendrites/plotting_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/plotting_utils.py]: The fix has been made here.

File: nupic/research/frameworks/dendrites/plotting_utils.py [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+    mask_values=None,
+    unit_to_plot=0
+):
+    """
+    Returns a heatmap of dendrite activations (given dendrite weights for a single
+    neuron and context vectors) plotted using matplotlib. Note that the user must be
+    logged in to wandb on both browser and terminal to view the resulting plots, and
+    this can be done via the following command:
+
+    $ wandb login your-login-key
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/plotting_utils.py]: Agreed; the statement has been removed from all docstrings.

File: nupic/research/frameworks/dendrites/utils.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py]: Yes, you're right. I have included that now.

File: nupic/research/frameworks/dendrites/utils.py (hunk 6 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py]: Fixed.

File: nupic/research/frameworks/dendrites/utils.py (hunk 4 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/utils.py]: I have changed the terminology in the files here to be consistent with you. In the future, I think we may need to propagate these changes to other dendrite files too.

Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/utils.py]: Thanks. I'll keep an eye out for where this occurs.

File: nupic/research/frameworks/dendrites/utils.py (hunk 5 above)
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/utils.py]: Much appreciated.

**Comment by mvacaporale** [comment]
> given just the 2 tensors `dendrite_activations` and `winning_mask` that you propose, I believe all metrics can still be computed.

Thanks for looking into this. I believe this may be the route to go. It's a significant change, so I'll leave it for a later PR. Beforehand, I'll explain my thinking moving forward.

1. These functions are general, but they still make some strict assumptions. The biggest assumption they make is how the winning dendrites are activated. Ideally, as we tweak this, these plotting functions should be immediately applicable. However, if we rely on them to calculate the winning segments then we'll have to revise as we go along. This is in the spirit of having [one job per function](https://en.wikipedia.org/wiki/Single-responsibility_principle#:~:text=The%20single%2Dresponsibility%20principle%20(SRP,functionality%2C%20which%20it%20should%20encapsulate.). Thus, our plotting and analysis functions should be only doing those things.

2. With constraint 1., only the dendrites segment object should be responsible for deciding the winning segments. Since we want these values, we'll need to do some plumbing. I'm thinking we can do an accompanying refactor that returns the winning indices (or possibly a mask), from the [apply-dendrites functions](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/dendrites/functional.py) which in turn can be returned by the DendriteSegements module. After that plumbing is in place, we can easily retrieve the winning indices/mask via a forward hook (which I'd recommend reading into if you haven't already).

**Review by mvacaporale — APPROVED** [review]

**Review by subutai — APPROVED** [review]

**Comment by subutai** [comment]
BTW, you should merge in the latest from master before merging.

File: nupic/research/frameworks/dendrites/plotting_utils.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/plotting_utils.py:37]: @subutai  The files here are not using wandb, and just plotting & returning a figure. That figure can then be either pasted to wandb or saved to disk. Since there's consideration for writing a mixin to handle logging of metrics, I propose we should leave saving files to that, and the functions here still return figures.

Comment (subutai) [reply nupic/research/frameworks/dendrites/plotting_utils.py:37]: Makes sense.
